designs/2020-es-module-support/README.md

```diff
@@ -0,0 +1,170 @@
+
+
+## Detailed Design
+
+### Nomenclature
+
+* `ESM` ecmascript module. The standardized module format for javascript
+* `CJS` commonjs. Node's historical module format (i.e., `module.exports = ...`, `const a = require('foo')`)
+* `UMD` universal module definition. A legacy format for specifying javascript code as a module in a way that tries to guarantee compatibility with many loader formats including requireJS, commonjs, and global script tags.
+* `IIFE` immediately invoked function expressions. A legacy strategy for encapsulating javascript into a module that won't leak global variables into the execution environment. It is still the primary way of packaging global script tags for inclusion in a web page. `UMD` is usually wrapped in an `IIFE`
+* `CDN` content delivery network. can be thought of as a set of servers that host static content all around the globe to deliver assets quickly and at scale.
+
+
+### build process
+
+Today, espree is written in commonjs (CJS) format, and a universal module definition (UMD) build is created too.
+
+
+today's build process:
+```
+┌-------------------┐               ┌-----------------┐
+│                   │               │                 │                package.json:
+│     espree.js     |               │ build/espree.js │
+│ (CJS entry point) ├──BUILD_STEP──▶│   (UMD bundle)  │     CJS ---▶   "main": "espree.js",
+│                   │               │                 │
+└-------------------┘               └-----------------┘
+```
+
+
+The strategy that probably makes the most sense in this initial work is to do what is known as a "dual package" solution, where
+espree provides an ESM, and a CJS module.
+
+Providing a dual package shouldn't affect any existing espree users, but provies an ESM option for users that want it.
+
+
+proposed build process in this RFC:
+```
+                                  ┌-------------------┐
+                                  │                   │
+                                  │  dist/espree.cjs  │
+                                  │ (CJS entry point) │                package.json:
+┌-------------------┐             │                   │
+│                   │             └--▲----------------┘     CJS ---▶   "main": "dist/espree.cjs",
+│     espree.js     │                │                                 "exports": {
+│ (ESM entry point) ├───BUILD_STEP───┘                      ESM ---▶     "import": "espree.js",
+│                   │                                       CJS ---▶     "require": "./dist/espree.cjs"
+└-------------------┘                                                  },
+                                                                       "type": "module"
+```
+
+browserify is specifically oriented around commonjs as the input format, so I also propose replacing it with rollup, which understands and expects ESM as it's input format.
```

Comment:
i'd be careful about this; rollup tends to violate the spec in various ways. browserify works just fine with ESM as far as i'm aware.

Reply:
We were previously only using browserify for a long-dead demo on the website. eslint/espree#466 removes it. What spec violations should we be on the lookout for with rollup?

Reply:
I don't have any current ones in mind; but for one, its entire heuristic for removing dead code is flawed - it will remove things that can throw exceptions, which broke es5-shim on airbnb.com for many months, which was relying on those exceptions being triggered. They fixed this by detecting the specific cases es5-shim is using, but that doesn't fix the root problem. For another, when `import * as` can't be statically detected with certainty that some imports are unused, rather than keeping them all or erroring, they opt to drop them, so if you're doing anything fancy with module namespace objects, things can break.

I'm sure issues will be addressed as they're found, but the overall approach and ideology of the project keeps me very wary.

Reply:
> I'm sure issues will be addressed as they're found, but the overall approach and ideology of the project keeps me very wary.

The only thing we're really using any bundler for is changing the `import` to `require` and `export` to `module.exports` for the cjs build.  I can't imagine there will be much spec violation if that's all we're doing?

>  browserify works just fine with ESM

huh, that's odd. My understanding is esm _still_ is unsupported in browserify. Was looking to see if this has landed and it's still on the wishlist:  https://github.com/browserify/browserify/issues/1186

Reply:
To clarify, it works with ESM in concert with babel.

Reply:
We'll bundle only our own code, so we probably don't need dead code removal. Can we turn it off with [treeshake: false](https://rollupjs.org/guide/en/#treeshake)?

Reply:
I'm pretty sure tree-shake is opt-in only; it's been experimental for as long as I know.

Reply:
I think the docs for the [treeshake option](https://rollupjs.org/guide/en/#treeshake) says `Default: true`.

Reply:
oh those crazy bastards enabled it by default now! :D ok, then yeah we should definitely shut off tree-shake.